Bluetooth: BR: Add listener cb for discovery

lylezhu2012 opened the pull request:
The results of inquiry and extended inquiry are only reported from application level after the inquiry complete event notified.

While the event of inquiry result is notified by controller in real time.

It is not a good user experience.

Just like scanning of LE, add a listener cb for discovery.

When the event of inquiry result, extended inquiry result, or remote name request complete notified, call listener `recv` cb to notify the upper layer.

When the event of inquiry complete notified or no pending of remote name request , call listener `timeout` cb to notify the upper layer.

lylezhu2012 committed:
Bluetooth: BR: Add listener cb for discovery

The results of inquiry and extended inquiry
are only reported from application level
after the inquiry complete event notified.

While the event of inquiry result is notified
by controller in real time.

It is not a good user experience.

Just like scanning of LE, add a listener
cb for discovery.

When the event of inquiry result, extended
inquiry result, or remote name request
complete notified, call listener `recv`
cb to notify the upper layer.

When the event of inquiry complete
notified or no pending of remote name
request , call listener `timeout`
cb to notify the upper layer.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jhedberg commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -2577,6 +2580,48 @@ int bt_br_discovery_start(const struct bt_br_discovery_param *param,
  */
 int bt_br_discovery_stop(void);
 
+struct bt_br_discovery_cb {
```
All these Classic-related APIs should probably be moved under `include/zephyr/bluetooth/classic/`

jhedberg replied:
..and I think it'd be good to do that before extending the current API

lylezhu2012 replied:
Updated

jhedberg commented:
> Just like scanning of LE, add a listener cb for discovery.

I *think* the reason why we have the two kind of scan callbacks for LE is due to backwards compatibility, i.e. the old singlelton callback was kept as-is when support for multiple independently registered callbacks was added. We don't necessarily have to follow that approach for BR/EDR since the API is not considered stable, i.e. let's do this the way we want to do it as if we were defining the API from scratch. @alwa-nordic @jori-nordic do you guys remember how this went for LE?

That said, I'm all in favor of making the API more asynchronous, i.e. the application should get updates as they happen, rather than the stack withholding the information until it's been able to get everything possible for each discovered device.

lylezhu2012 committed:
Bluetooth: classic: Move classic source to sub folder

Move source code of classic from bluetooth.h to
./classic/classic.h.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
Bluetooth: Classic: Remove callback from bt_br_discovery_start

Remove BR discovery callback from bt_br_discovery_start.

All discovery results will be notified through callback
registered by bt_br_discovery_cb_register.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 committed:
sample: Bluetooth: hfp_ag: update device discovery

Due to the update of function bt_br_discovery_start,
register discovery callback by calling
bt_br_discovery_cb_register.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented:
> > Just like scanning of LE, add a listener cb for discovery.
> 
> I _think_ the reason why we have the two kind of scan callbacks for LE is due to backwards compatibility, i.e. the old singlelton callback was kept as-is when support for multiple independently registered callbacks was added. We don't necessarily have to follow that approach for BR/EDR since the API is not considered stable, i.e. let's do this the way we want to do it as if we were defining the API from scratch. @alwa-nordic @jori-nordic do you guys remember how this went for LE?
> 
> That said, I'm all in favor of making the API more asynchronous, i.e. the application should get updates as they happen, rather than the stack withholding the information until it's been able to get everything possible for each discovered device.

Well. I removed the callback parameter from function bt_br_discovery_start.

lylezhu2012 committed:
Bluetooth: shell: BR: update device discovery

Due to the update of function bt_br_discovery_start,
register discovery callback by calling
bt_br_discovery_cb_register.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>
